src/sentry/utils/circuit_breaker2.py

```diff
@@ -0,0 +1,185 @@
+"""
+NOTE: This is a replacement for the current circuit breaker implementation, which is why it is
+`circuit_breaker2`. It's first going to be used for the Seer similarity service, then once we're
+confident it works we can replace use of the original for the severity service with use of this one
+and get rid of the old one, at which point this can lose the `2`.
+"""
+
+import logging
+from enum import Enum
+from typing import NotRequired, TypedDict
+
+from sentry.ratelimits.sliding_windows import Quota, RedisSlidingWindowRateLimiter
+
+logger = logging.getLogger(__name__)
+
+# XXX: If either of these values changes, the `CircuitBreakerConfig` docstrings
+# need to be updated
+# How many times stricter to be with the error limit during recovery
+DEFAULT_RECOVERY_STRICTNESS = 10
+# How many times the length of the error window to make the recovery window
+DEFAULT_RECOVERY_WINDOW_MULTIPLIER = 2
+
+
+class CircuitBreakerState(Enum):
+    OK = "circuit_okay"
+    BROKEN = "circuit_broken"
+    RECOVERY = "recovery"
+
+
+class CircuitBreakerConfig(TypedDict):
+    # The number of errors within the given time period necessary to trip the breaker
+    error_limit: int
+    # The time period, in seconds, over which we're tracking errors
+    error_limit_window: int
+    # How long, in seconds, to stay in the BROKEN state (blocking all requests) before entering the
+    # RECOVERY phase
+    broken_state_duration: int
+    # The number of errors within the given time period necessary to trip the breaker while in
+    # RECOVERY. Will be set automatically to 10% of `error_limit` if not provided.
+    recovery_error_limit: NotRequired[int]
+    # The length, in seconds, of each time bucket ("granule") used by the underlying rate limiter -
+    # effectively the resolution of the time window. Will be set automatically based on
+    # `error_limit_window` if not provided.
+    error_limit_window_granularity: NotRequired[int]
+    # How long, in seconds, to stay in the RECOVERY state (allowing requests but with a stricter
+    # error limit) before returning to normal operation. Will be set to twice `error_limit_window`
+    # if not provided.
+    recovery_duration: NotRequired[int]
+
+
+# XXX: This limit is intentionally very high, because without active configuration to override these
+# values, we'd rather have the circuit breaker not engage when maybe it should (we also have
+# killswitches, rate limiters, feature flags, and the like upon which we can rely to shut things
+# down) than have it engage prematurely.
+#
+# TODO: Consider making this percentage-based rather than having an absolute limit
+CIRCUIT_BREAKER_DEFAULT_CONFIG: CircuitBreakerConfig = {
+    "error_limit": 11211231,
+    "error_limit_window": 3600,  # 1 hr
+    "broken_state_duration": 300,  # 5 min
+}
```

Comment:
My theory was that it's both. (See [here](https://github.com/getsentry/sentry/pull/74563/files#diff-c4bf5bc2eae9efe5e948b5deda98c34fafd95674ae376aaf790c21a039e59f91R895-R909) where the defaults for Seer similarity are set.) My thinking was that if the config were a dataclass rather than a typeddict, I'd've included default values, and so this basically doing that. I guess there's no reason we couldn't have the default-defaults as well as the service-specific defaults in options, though. I'll make the change.

Reply:
UPDATE: In the end I decided that the best option was not to have a default at all, so that if you're going to use the circuit breaker, you have to at least think about what numbers you want.